src/util/schnorr.rs

```diff
@@ -18,64 +18,114 @@
 //!
 
 pub use secp256k1::schnorrsig::{PublicKey, KeyPair};
-use secp256k1::{Secp256k1, Verification};
+use secp256k1::{Secp256k1, Verification, Error, constants, Signing};
 use hashes::{Hash, HashEngine};
 use util::taproot::{TapBranchHash, TapTweakHash};
+use core::{fmt, str};
 
 /// Untweaked Schnorr public key
 pub type UntweakedPublicKey = PublicKey;
 
 /// Tweaked Schnorr public key
+#[derive(Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash, Debug)]
 pub struct TweakedPublicKey(PublicKey);
 
+impl fmt::LowerHex for TweakedPublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(&self.0, f)
+    }
+}
+
+impl fmt::Display for TweakedPublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(&self.0, f)
+    }
+}
+
+impl str::FromStr for TweakedPublicKey {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<TweakedPublicKey, Error> {
+        Ok(TweakedPublicKey(PublicKey::from_str(s)?))
+    }
+}
+
+/// Tweaked BIP-340 key pair
+#[derive(Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash, Debug)]
+pub struct TweakedKeyPair(KeyPair);
+
 /// A trait for tweaking Schnorr public keys
-pub trait TapTweak {
-    /// Tweaks an untweaked public key given an untweaked key and optional script tree merkle root.
-    ///
-    /// This is done by using the equation Q = P + H(P|c)G, where
-    ///  * Q is the tweaked key
-    ///  * P is the internal key
+pub trait TapTweak<U, T> {
```

Comment:
Documentation on what `T` and `U` are would be helpful. Preferably, *also* use words instead of letters for non-obvious generics.

Also, maybe this trait should be called `Secp256k1Ext` and/or sealed?

Reply:
This belongs to a different PR - will address it there